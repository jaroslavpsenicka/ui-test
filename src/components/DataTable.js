import React, { useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { getData } from '../redux/actions'

const INTERVAL = process.env.REACT_APP_REFRESH_INTERVAL || 10000

const DataTable = () => {

  const dispatch = useDispatch()
  const data = useSelector(state => state.data)

  useEffect(() => {
    dispatch(getData())
    const interval = setInterval(() => dispatch(getData()), INTERVAL)
    return () => clearInterval(interval)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const Table = ({ data }) => (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Height</th>
        </tr>
      </thead>
      <tbody>
      { 
        data.sort((a, b) => a.name.localeCompare(b.name))
          .map((item, i) => <tr key={`${item.name}-${item.height}-${i}`}>
            <td>{item.name}</td>
            <td>{item.height}</td>
          </tr>)     
      }
      </tbody>
    </table>
  )

  return data?.success ? <Table data={data.data} /> : <div>{data.error}</div>    
}

export default DataTable