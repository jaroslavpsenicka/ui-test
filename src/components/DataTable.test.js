import React from 'react';
import { Provider } from 'react-redux'
import { mount, shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import toJson from 'enzyme-to-json';

import DataTable from './DataTable';

const mockStore = configureMockStore([thunk])

describe('DataTable', () => {

  it('renders no data', () => {
    const store = mockStore({ data: {} })
    const wrapper = mount(<Provider store={store}><DataTable /></Provider>)
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders error', () => {
    const store = mockStore({ data: { success: false, error: 'ERR' } })
    const wrapper = mount(<Provider store={store}><DataTable /></Provider>)
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders one row', () => {
    const store = mockStore({ data: { success: true, data: [{ name: 'A', height: 5 }]}})
    const wrapper = mount(<Provider store={store}><DataTable /></Provider>)
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders two rows, sorts properly', () => {
    const store = mockStore({ data: { success: true, data: [{ name: 'B', height: 2 }, { name: 'A', height: 5 }]}})
    const wrapper = mount(<Provider store={store}><DataTable /></Provider>)
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
