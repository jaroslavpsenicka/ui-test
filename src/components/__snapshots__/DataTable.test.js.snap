// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders error 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DataTable>
    <div>
      ERR
    </div>
  </DataTable>
</Provider>
`;

exports[`App renders no data 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DataTable>
    <div />
  </DataTable>
</Provider>
`;

exports[`App renders one row 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DataTable>
    <Table
      data={
        Array [
          Object {
            "height": 5,
            "name": "A",
          },
        ]
      }
    >
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Height
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="A-5-0"
          >
            <td>
              A
            </td>
            <td>
              5
            </td>
          </tr>
        </tbody>
      </table>
    </Table>
  </DataTable>
</Provider>
`;

exports[`App renders two rows, sorts properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DataTable>
    <Table
      data={
        Array [
          Object {
            "height": 5,
            "name": "A",
          },
          Object {
            "height": 2,
            "name": "B",
          },
        ]
      }
    >
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Height
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="A-5-0"
          >
            <td>
              A
            </td>
            <td>
              5
            </td>
          </tr>
          <tr
            key="B-2-1"
          >
            <td>
              B
            </td>
            <td>
              2
            </td>
          </tr>
        </tbody>
      </table>
    </Table>
  </DataTable>
</Provider>
`;
